tools/translate.sh

```diff
@@ -0,0 +1,138 @@
+#!/bin/sh
+# Shell script helper for translating Metamath tool files to bash
+# Mario Carneiro 7-Jan-2022
+
+usage() {
+  cat << HELP
+TRANSLATE - Translate a TOOLS script into a corresponding bash script
+
+This command will take a script file intended for use with metamath
+tools-mode, and translate it into a bash script with the same behavior.
+See also SUBMIT. The script is passed the location of the tools
+directory (containing the core script files), which is also the location
+of this file. It uses 'mm-tools' by default, so if you symlink 'mm-tools'
+to the tools directory then these scripts will work.
+
+If the -r (recursive) option is passed, the generated file will translate
+SUBMIT calls to also make use of the same TOOLS directory; otherwise
+it will leave them off (using the default "mm-tools" directory).
+
+If the -e (echo) option is passed, the generated file will contain an
+echo statement for each executed command.
+
+Syntax:  translate.sh [-r] [-e] [TOOLS] < INFILE > OUTFILE
+HELP
+}
+
+if [ $# -eq 1 ] && [ "$1" = "-h" -o "$1" = "--help" ]; then usage; exit; fi
+
+if [ "$1" = "-r" ]; then recursive=1; shift; fi
+if [ "$1" = "-e" ]; then echo=1; shift; fi
+
+if [ $# -eq 0 ]; then tools="mm-tools"
+elif [ $# -eq 1 ]; then tools=$1
+else usage; exit 1; fi
+
+echo "#!/bin/bash"
+awk -v tools="$tools" -v echo="$echo" -v recursive="$recursive" \
+'BEGIN {
+  len = length(specChars);
+  split(specChars, specCharsArr, "");
+  if (recursive) {
+    extraArg = tools;
+    gsub(/[\\$"]/, "\\\\&", extraArg);
```

Comment:
according to the bash manual, the apostrophe ' needs quoting as well.

Reply:
You sure? `echo "I'm testing quotes"` seems to work just fine

Reply:
From the bash manual:
3.1.2.3 Double Quotes
Enclosing characters in double quotes (‘"’) preserves the literal value of all characters within
the quotes, with the exception of ‘$’, ‘‘’, ‘\’, and, when history expansion is enabled, ‘!’.

Reply:
Although I see that text in the manual, I also can't find any special meaning for `'` outside of its interaction with `${...}` which doesn't matter if all `$` are quoted. The history expansion thing looks worrisome but based on my testing it's not active when reading from scripts.

Reply:
Oh! That character isn't the apostrophe, it's a backtick. ``"`command`"`` can be used to execute a command and insert the result into the string, and indeed it needs to be escaped.

Reply:
line 64 is also affected